javaslang/src/main/java/javaslang/collection/Vector.java

```diff
@@ -1387,23 +1414,376 @@ public String toString() {
         return mkString(stringPrefix() + "(", ", ", ")");
     }
 
-    private static <T> Vector<T> wrap(HashArrayMappedTrie<Integer, T> trie) {
-        return trie.isEmpty() ? empty() : new Vector<>(trie);
+    enum Helper {
+        ;
+        static <T> VectorTree<T> build(Object[] array, int offset, int length, int depth) {
+            assert (array != null) && (offset >= 0) && (length >= 0);
+            if (array.length == 0) {
+                return emptyTree();
+            } else if (depth == 1) {
+                return new Vector1<>((T[]) array);
+            } else {
+                return new VectorN<>(array, offset, length, depth);
+            }
+        }
+
+        static <T> T[] getLeaf(Object[] root, int depth, int index) {
+            assert (depth >= 2) && (index >= 0);
+
+            int shift = shiftValue(depth);
+            Object[] array = (Object[]) root[index >> shift]; // doesn't need masking
+
+            while (shift > BRANCHING_BASE) {
+                shift -= BRANCHING_BASE;
+                array = (Object[]) array[(index >> shift) & branchingMask()]; // TODO mask and shift should be hidden in a proper method, e.g. nth
+            }
+
+            assert array != null;
+            return (T[]) array;
+        }
+
+        static int shiftValue(int depth)   { return (depth - 1) * BRANCHING_BASE; }
+        static int fullTreeSize(int depth) { return branchingFactor() * fullBranchSize(depth); }
+
+        /* branchingFactor()^(depth-1) == 2^(BRANCHING_BASE*depth) */
+        static int fullBranchSize(int depth) {
+            final int result = 1 << (BRANCHING_BASE * (depth - 1));
+            assert result == (int) Math.pow(branchingFactor(), (depth - 1));
+            return result;
+        }
     }
-}
 
-interface VectorModule {
+    static final class Leaf<T> implements Serializable { // TODO separate start/end?
+        private static final long serialVersionUID = 1L;
+        private static final Leaf<?> EMPTY = new Leaf<>(emptyArray(), (byte) 0);
 
-    final class Combinations {
+        private final T[] array;
+        private final byte length;
 
-        static <T> Vector<Vector<T>> apply(Vector<T> elements, int k) {
-            if (k == 0) {
-                return Vector.of(Vector.empty());
+        static <T> Leaf<T> empty()                         { return (Leaf<T>) EMPTY; }
+        public static <T> Leaf<T> of(T[] leaf)             { return of(leaf, leaf.length); }
+        public static <T> Leaf<T> of(T[] leaf, int length) { return new Leaf<T>(leaf, (byte) length); }
+
+        private Leaf(T[] array, byte length) {
```

Comment:
This enables fast `drop` for front and back, i.e. in case of the leading and trailing arrays (which are not part of the tree structure), we can keep the full array and store how much we've dropped.
This is useful in case of multiple drop calls (e.g. `tail`) and has a constant memory surplus (in worst case, `2*(branchingFactor-1)`)